service/src/Helper.cpp

```diff
@@ -269,6 +271,46 @@ namespace geopm
         return stat_struct.st_gid;
     };
 
+    unsigned int fd_to_uid(const int fd) {
+        int err = 0;
+        struct stat stat_struct;
+        err = fstat(fd, &stat_struct);
+        if (err) {
+            throw Exception("fd_to_uid(): ", errno ? errno : GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+        }
+        return stat_struct.st_uid;
+    }
+
+    unsigned int fd_to_mode(const int fd) {
+        int err = 0;
+        struct stat stat_struct;
+        err = fstat(fd, &stat_struct);
+        if (err) {
+            throw Exception("fd_to_mode(): ", errno ? errno : GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+        }
+        return stat_struct.st_mode;
+    }
+
+    bool is_plugin_secure(const int fd, const std::string &plugin) {
+        bool is_secure = true;
+        if (getuid() == 0) {
+            int uid = geopm::fd_to_uid(fd);
+            if (uid != 0) {
+                is_secure = false;
+                std::cerr << "Warning: <geopm> Plugin not owned by root and will be ignored: "
+                          << plugin << std::endl;
+            }
+            int mode = geopm::fd_to_mode(fd);
+            if (mode != (S_IFREG | S_IRWXU)) { // 0o700
```

Comment:
Default plugins should have permissions 755 like other files in /usr/lib64.  So the validation here must be that the write bits are not set. For example 755 is okay, 700 is okay, but 777 is not okay